src/ActionServer_FJT.c

```diff
@@ -495,18 +495,59 @@ void Ros_ActionServer_FJT_Goal_Complete(GOAL_END_TYPE goal_end_type)
         {
-            diff = fabs(feedback_FollowJointTrajectory.feedback.desired.positions.data[axis] - feedback_FollowJointTrajectory.feedback.actual.positions.data[axis]);
+            rosidl_runtime_c__String selected_tolerance_name = g_actionServer_FJT_SendGoal_Request.goal.goal_tolerance.data[selected_tolerance_idx].name;
+            Ros_Debug_BroadcastMsg("FJT found position tolerance: %s, %f", selected_tolerance_name, g_actionServer_FJT_SendGoal_Request.goal.goal_tolerance.data[selected_tolerance_idx].position);
 
-            double posTolerance = g_actionServer_FJT_SendGoal_Request.goal.goal_tolerance.data[axis].position; //user-provided a goal_tolerance
-            if (posTolerance == 0.0) //user did NOT provide a tolerance
-                posTolerance = DEFAULT_FJT_GOAL_POSITION_TOLERANCE;
+            for (size_t selected_position_idx = 0; selected_position_idx < feedback_FollowJointTrajectory.feedback.joint_names.size; selected_position_idx += 1)
+            {
+                rosidl_runtime_c__String selected_position_name = feedback_FollowJointTrajectory.feedback.joint_names.data[selected_position_idx];
+                if (rosidl_runtime_c__String__are_equal(&selected_position_name,&selected_tolerance_name))
+                {
+                    Ros_Debug_BroadcastMsg("FJT using position tolerance: %s", selected_position_name);
+                    posTolerance[selected_position_idx] = g_actionServer_FJT_SendGoal_Request.goal.goal_tolerance.data[selected_tolerance_idx].position;
+                    break;
+                }
+            }
+        }
 
-            if (diff > posTolerance)
+        double finalPositions[maxAxes];
+        bzero(finalPositions, sizeof(finalPositions));
+        size_t trajectorySizeOffset = g_actionServer_FJT_SendGoal_Request.goal.trajectory.points.size - 1;
+        for (size_t selected_final_position_idx = 0; selected_final_position_idx < g_actionServer_FJT_SendGoal_Request.goal.trajectory.joint_names.size; selected_final_position_idx += 1)
+        {
+            rosidl_runtime_c__String selected_final_position_name = g_actionServer_FJT_SendGoal_Request.goal.trajectory.joint_names.data[selected_final_position_idx];
+            for (size_t selected_position_idx = 0; selected_position_idx < feedback_FollowJointTrajectory.feedback.joint_names.size; selected_position_idx += 1)
+            {
+                rosidl_runtime_c__String selected_position_name = feedback_FollowJointTrajectory.feedback.joint_names.data[selected_position_idx];
+                if (rosidl_runtime_c__String__are_equal(&selected_position_name, &selected_final_position_name))
+                {
+                    finalPositions[selected_position_idx] = g_actionServer_FJT_SendGoal_Request.goal.trajectory.points.data[trajectorySizeOffset].positions.data[selected_position_idx];
```

Comment:
I haven't worked it out in detail, but should this be:

```suggestion
                    finalPositions[selected_position_idx] = g_actionServer_FJT_SendGoal_Request.goal.trajectory.points.data[trajectorySizeOffset].positions.data[selected_final_position_index];
```

?

Reply:
You are correct, during my testing my final positions were all the same, so I didn't run into an issue with this line.